ledger/accountdb.go

```diff
@@ -1026,74 +1080,395 @@ func accountsPutTotals(tx *sql.Tx, totals ledgercore.AccountTotals, catchpointSt
+	// 3. if old is above the threshold
+	//  - find created and deleted entries
+	//  - if the result is below the threshold then move everything in Assets field
+	//  - if the result is above the threshold then merge in changes into group data
+	var pad ledgercore.PersistedAccountData
+	var err error
+	if delta.old.pad.NumAssetHoldings() <= assetsThreshold && len(delta.new.Assets) <= assetsThreshold {
+		pad.AccountData = delta.new
+	} else if delta.old.pad.NumAssetHoldings() <= assetsThreshold && len(delta.new.Assets) > assetsThreshold {
+		pad.ExtendedAssetHolding.ConvertToGroups(delta.new.Assets)
+		pad.AccountData = delta.new
+		pad.AccountData.Assets = nil
+
+		// update group data DB table
+		var result sql.Result
+		for i := 0; i < len(pad.ExtendedAssetHolding.Groups); i++ {
+			result, err = qaei.Exec(pad.ExtendedAssetHolding.Groups[i].Encode())
+			if err != nil {
+				break
+			}
+			pad.ExtendedAssetHolding.Groups[i].AssetGroupKey, err = result.LastInsertId()
+			if err != nil {
+				break
+			}
+		}
+	} else { // default case: delta.old.pad.NumAssetHoldings() > assetsThreshold
+		deleted := make([]basics.AssetIndex, 0, len(delta.old.pad.Assets)) // items in old but not in new
+		updated := make([]basics.AssetIndex, 0, len(delta.new.Assets))     // items in both new and old
+		created := make([]basics.AssetIndex, 0, len(delta.new.Assets))     // items in new but not in old
+
+		pad.AccountData = delta.new
+		oldPad := delta.old.pad
+
+		for aidx := range oldPad.AccountData.Assets {
+			if _, ok := delta.new.Assets[aidx]; ok {
+				updated = append(updated, aidx)
+			} else {
+				deleted = append(deleted, aidx)
+			}
+		}
+		for aidx := range delta.new.Assets {
+			if _, ok := oldPad.AccountData.Assets[aidx]; ok {
+				// updated = append(updated, aidx)
+				// handled above as updated
+			} else {
+				created = append(created, aidx)
+			}
+		}
+		newCount := oldPad.NumAssetHoldings() + len(created) - len(deleted)
+		if newCount < assetsThreshold {
+			// Move all assets from groups to Assets field
+			assets, _, err := loadHoldings(qabq, oldPad.ExtendedAssetHolding)
+			if err != nil {
+				return updatedAccounts, err
+			}
+			for _, aidx := range deleted {
+				delete(assets, aidx)
+			}
+			for _, aidx := range created {
+				assets[aidx] = delta.new.Assets[aidx]
+			}
+			for _, aidx := range updated {
+				assets[aidx] = delta.new.Assets[aidx]
+			}
+			pad.AccountData.Assets = assets
+
+			// now delete all old groups
+			for _, g := range oldPad.ExtendedAssetHolding.Groups {
+				qaed.Exec(g.AssetGroupKey)
+			}
+			pad.ExtendedAssetHolding = ledgercore.ExtendedAssetHolding{}
+		} else {
+			// Reconcile groups data:
+			// identify groups, load, update, then delete and insert, dump to the disk
+
+			pad.ExtendedAssetHolding = oldPad.ExtendedAssetHolding
+			if len(updated) > 0 {
+				sort.SliceStable(updated, func(i, j int) bool { return updated[i] < updated[j] })
+				gi, ai := 0, 0
+				for _, aidx := range updated {
```

Comment:
this entire block re-written after introducing `delta.holding` that track deletion in next PR with tests. Please skip with branch for now